README.rst
.. image:: https://img.shields.io/badge/chat-join%20now-blue.svg
   :target: https://gitter.im/python-trio/general
   :alt: Join chatroom

.. image:: https://img.shields.io/badge/docs-read%20now-blue.svg
   :target: https://asyncowfs.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation Status
   
.. image:: https://img.shields.io/pypi/v/asyncowfs.svg
   :target: https://pypi.org/project/asyncowfs
   :alt: Latest PyPi version

.. image:: https://travis-ci.org/python-trio/asyncowfs.svg?branch=master
   :target: https://travis-ci.org/python-trio/asyncowfs
   :alt: Automated test status

.. image:: https://codecov.io/gh/python-trio/asyncowfs/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/python-trio/asyncowfs
   :alt: Test coverage


Async 1wire
===========

Welcome to `Async 1wire <https://github.com/smurfix/asyncowfs>`__!

This library allows you to access 1wire buses with Python, using any
async library (trio, asyncio, curio).

AsyncOWFS does *not* try to talk to any 1wire interface by itself; it
uses TCP port 4304 (by default) to access one or more ``owserver``\ s.

Bus enumeration (when starting up, plus periodically in the background
if desired) is automatic and can be turned off.

This module supports background polling, alerts, and simultaneous
conversions. Bus couplers (DS2509) are handled transparently. Accessors to
device attributes are created from device descriptors, read from
``owserver``, and augmented by device-specific classes when necessary.

AsyncOWFS includes a test suite, including a mock 1wire server written in
Python, which you can easily re-use to test your own code. See `distowfs
<https://github.com/smurfix/distowfs>`__ for an example.

License: Your choice of MIT or Apache License 2.0



debian/changelog
asyncowfs (0.11.1-1) unstable; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 30 Jun 2020 09:02:59 +0200

asyncowfs (0.11.0-1) unstable; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 30 Jun 2020 04:19:45 +0200

asyncowfs (0.10.3-1) unstable; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 29 Jun 2020 12:54:49 +0200

asyncowfs (0.10.2-1) unstable; urgency=medium

  * Merge 0.10.2

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 10 Aug 2019 19:08:29 +0200

asyncowfs (0.10.1-2) unstable; urgency=medium

  * Packaging woes

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 10 Aug 2019 10:43:28 +0200

asyncowfs (0.10.1-1) unstable; urgency=medium

  * Merge 0.10.1

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 10 Aug 2019 10:40:15 +0200

asyncowfs (0.10.0-1) unstable; urgency=medium

  * Merge 0.10.0

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 07 Aug 2019 12:02:12 +0200

asyncowfs (0.9.5-1) unstable; urgency=medium

  * Merge 0.9.5

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 30 Jul 2019 19:03:55 +0200

asyncowfs (0.9.3-1) unstable; urgency=medium

  * Merge 0.9.3

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 30 Jul 2019 11:58:24 +0200

asyncowfs (0.9.1-2) unstable; urgency=medium

  * Rename

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 29 Jul 2019 21:27:49 +0200

trio-owfs (0.7.0-1) smurf; urgency=medium

  * New release

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 12 Nov 2018 09:34:11 +0100

trio-owfs (0.6.7-1) smurf; urgency=medium

  * New release

 -- Matthias Urlichs <matthias@urlichs.de>  Fri, 02 Nov 2018 12:21:24 +0100

trio-owfs (0.6.5-1) smurf; urgency=medium

  * Version uptick

 -- Matthias Urlichs <matthias@urlichs.de>  Fri, 26 Oct 2018 13:39:44 +0200

trio-owfs (0.5.0-1) smurf; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 24 Oct 2018 14:30:18 +0200

trio-owfs (0.4.2-1) smurf; urgency=low

  * source package automatically created by stdeb 0.8.5

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 14 Oct 2018 08:58:52 +0200


asyncowfs/error.py
# for message type classification see
# http://owfs.org/index.php?page=owserver-message-types
# and 'enum msg_classification' from module/owlib/src/include/ow_message.h

import attr

import logging
logger = logging.getLogger(__name__)


class OWFSReplyError(Exception):
    pass


@attr.s(eq=False)
class GenericOWFSReplyError(OWFSReplyError):
    err = attr.ib()
    req = attr.ib()
    server = attr.ib()


_errors = {}


@attr.s(eq=False)
class OWFSReplyError_(OWFSReplyError):
    req = attr.ib()
    server = attr.ib()


def _register(cls):
    _errors[cls.err] = cls
    return cls


@_register
@attr.s(eq=False)
class PermissionError(OWFSReplyError_):
    err = 13


@_register
@attr.s(eq=False)
class InUseError(OWFSReplyError_):
    err = 98


@_register
@attr.s(eq=False)
class NotAvailableError(OWFSReplyError_):
    err = 99


@_register
@attr.s(eq=False)
class TryAgainError(OWFSReplyError_):
    err = 11


@_register
@attr.s(eq=False)
class BadFSError(OWFSReplyError_):
    err = 9


@_register
@attr.s(eq=False)
class BadMsgError(OWFSReplyError_):
    err = 74


@_register
@attr.s(eq=False)
class BusyError(OWFSReplyError_):
    err = 16


@_register
@attr.s(eq=False)
class ConnAbortedError(OWFSReplyError_):
    err = 103


@_register
@attr.s(eq=False)
class FaultError(OWFSReplyError_):
    err = 14


@_register
@attr.s(eq=False)
class InterruptedError(OWFSReplyError_):
    err = 4


@_register
@attr.s(eq=False)
class InvalidDataError(OWFSReplyError_):
    err = 22


@_register
@attr.s(eq=False)
class BusIOError(OWFSReplyError_):
    err = 5


@_register
@attr.s(eq=False)
class IsDirError(OWFSReplyError_):
    err = 21


@_register
@attr.s(eq=False)
class LoopError(OWFSReplyError_):
    err = 40


@_register
@attr.s(eq=False)
class MsgSizeError(OWFSReplyError_):
    err = 90


@_register
@attr.s(eq=False)
class NameTooLongError(OWFSReplyError_):
    err = 36


@_register
@attr.s(eq=False)
class NoBufsError(OWFSReplyError_):
    err = 105


@_register
@attr.s(eq=False)
class NoDeviceError(OWFSReplyError_):
    err = 19


@_register
@attr.s(eq=False)
class NoEntryError(OWFSReplyError_):
    err = 2


@_register
@attr.s(eq=False)
class NoFreeMemoryError(OWFSReplyError_):
    err = 12


@_register
@attr.s(eq=False)
class NoMessageError(OWFSReplyError_):
    err = 42


@_register
@attr.s(eq=False)
class NoDirectoryError(OWFSReplyError_):
    err = 20


@_register
@attr.s(eq=False)
class NotSupportedError(OWFSReplyError_):
    err = 95


@_register
@attr.s(eq=False)
class RangeError(OWFSReplyError_):
    err = 34


@_register
@attr.s(eq=False)
class ReadOnlyError(OWFSReplyError_):
    err = 30


@_register
@attr.s(eq=False)
class InputPathTooLongError(OWFSReplyError_):
    err = 26


@_register
@attr.s(eq=False)
class BadPathSyntaxError(OWFSReplyError_):
    err = 27


@_register
@attr.s(eq=False)
class TextInPathError(OWFSReplyError_):
    err = 77


@_register
@attr.s(eq=False)
class BadCRC8Error(OWFSReplyError_):
    err = 28


@_register
@attr.s(eq=False)
class UnknownNameError(OWFSReplyError_):
    err = 29


@_register
@attr.s(eq=False)
class AliasTooLongError(OWFSReplyError_):
    err = 31


@_register
@attr.s(eq=False)
class UnknownPropertyError(OWFSReplyError_):
    err = 32


@_register
@attr.s(eq=False)
class NotAnArrayError(OWFSReplyError_):
    err = 33


@_register
@attr.s(eq=False)
class IsAnArrayError(OWFSReplyError_):
    err = 35


@_register
@attr.s(eq=False)
class NotBitfieldError(OWFSReplyError_):
    err = 37


@_register
@attr.s(eq=False)
class IndexTooLargeError(OWFSReplyError_):
    err = 38


@_register
@attr.s(eq=False)
class NoSubpathError(OWFSReplyError_):
    err = 39


@_register
@attr.s(eq=False)
class DeviceNotFoundError(OWFSReplyError_):
    err = 41


@_register
@attr.s(eq=False)
class DeviceError(OWFSReplyError_):
    err = 43


@_register
@attr.s(eq=False)
class BusShortError(OWFSReplyError_):
    err = 44


@_register
@attr.s(eq=False)
class NoSuchBusError(OWFSReplyError_):
    err = 45


@_register
@attr.s(eq=False)
class BusNotAppropriateError(OWFSReplyError_):
    err = 46


@_register
@attr.s(eq=False)
class BusNotRespondingError(OWFSReplyError_):
    err = 47


@_register
@attr.s(eq=False)
class BusResetError(OWFSReplyError_):
    err = 48


@_register
@attr.s(eq=False)
class BusClosedError(OWFSReplyError_):
    err = 49


@_register
@attr.s(eq=False)
class BusNotOpenedError(OWFSReplyError_):
    err = 50


@_register
@attr.s(eq=False)
class BusCommunicationError(OWFSReplyError_):
    err = 51


@_register
@attr.s(eq=False)
class BusTimeoutError(OWFSReplyError_):
    err = 52


@_register
@attr.s(eq=False)
class TelnetError(OWFSReplyError_):
    err = 53


@_register
@attr.s(eq=False)
class TCPError(OWFSReplyError_):
    err = 54


@_register
@attr.s(eq=False)
class BusIsLocalError(OWFSReplyError_):
    err = 55


@_register
@attr.s(eq=False)
class BusIsRemoteError(OWFSReplyError_):
    err = 56


@_register
@attr.s(eq=False)
class ReadTooLargeError(OWFSReplyError_):
    err = 57


@_register
@attr.s(eq=False)
class DataCommunicationError(OWFSReplyError_):
    err = 58


@_register
@attr.s(eq=False)
class NotRPropertyError(OWFSReplyError_):
    err = 59


@_register
@attr.s(eq=False)
class NotReadablePropertyError(OWFSReplyError_):
    err = 60


@_register
@attr.s(eq=False)
class DataTooLargeError(OWFSReplyError_):
    err = 61


@_register
@attr.s(eq=False)
class DataTooSmallError(OWFSReplyError_):
    err = 62


@_register
@attr.s(eq=False)
class DataFormatError(OWFSReplyError_):
    err = 63


@_register
@attr.s(eq=False)
class NotWPropertyError(OWFSReplyError_):
    err = 64


@_register
@attr.s(eq=False)
class NotWritablePropertyError(OWFSReplyError_):
    err = 65


@_register
@attr.s(eq=False)
class ReadOnlyModeError(OWFSReplyError_):
    err = 66


@_register
@attr.s(eq=False)
class DataCommError(OWFSReplyError_):
    err = 67


@_register
@attr.s(eq=False)
class OutputPathTooLongError(OWFSReplyError_):
    err = 68


@_register
@attr.s(eq=False)
class NotADirectoryError(OWFSReplyError_):
    err = 69


@_register
@attr.s(eq=False)
class NotADeviceError(OWFSReplyError_):
    err = 70


@_register
@attr.s(eq=False)
class UnknownQueryError(OWFSReplyError_):
    err = 71


@_register
@attr.s(eq=False)
class SocketError(OWFSReplyError_):
    err = 72


@_register
@attr.s(eq=False)
class TimeoutError(OWFSReplyError_):
    err = 73


@_register
@attr.s(eq=False)
class VersionError(OWFSReplyError_):
    err = 75


@_register
@attr.s(eq=False)
class PacketSizeError(OWFSReplyError_):
    err = 76


@_register
@attr.s(eq=False)
class UnexpectedNullError(OWFSReplyError_):
    err = 78


@_register
@attr.s(eq=False)
class NoMemoryError(OWFSReplyError_):
    err = 79
